x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1039,32 +1055,32 @@ private static Expression processBinaryOperator(BinaryOperator<?, ?, ?, ?> o) {
             }
             List<Expression> newChildren = new ArrayList<>(2);
             boolean childrenChanged = false;
-            DataType targetDataType = DataType.NULL;
-            Expression from = Literal.NULL;
-
-            if (left.dataType() == KEYWORD && left.foldable() && (left instanceof EsqlScalarFunction == false)) {
-                if (supportsStringImplicitCasting(right.dataType())) {
-                    targetDataType = right.dataType();
-                    from = left;
-                } else if (supportsImplicitTemporalCasting(right, o)) {
-                    targetDataType = DATETIME;
-                    from = left;
+            String errorMessage = "Cannot convert string [{}] to any of [" + DATETIME + ", " + DATE_PERIOD + ", " + TIME_DURATION + "]";
+
+            if (o instanceof DateTimeArithmeticOperation) {
+                childrenChanged = processDateTimeArithmeticOperation(left, right, newChildren, errorMessage);
+            } else {
+                DataType targetDataType = DataType.NULL;
+                Expression from = Literal.NULL;
+
+                if (left.dataType() == KEYWORD && left.foldable() && (left instanceof EsqlScalarFunction == false)) {
+                    if (supportsStringImplicitCasting(right.dataType())) {
+                        targetDataType = right.dataType();
+                        from = left;
+                    }
                 }
-            }
-            if (right.dataType() == KEYWORD && right.foldable() && (right instanceof EsqlScalarFunction == false)) {
-                if (supportsStringImplicitCasting(left.dataType())) {
-                    targetDataType = left.dataType();
-                    from = right;
-                } else if (supportsImplicitTemporalCasting(left, o)) {
-                    targetDataType = DATETIME;
-                    from = right;
+                if (right.dataType() == KEYWORD && right.foldable() && (right instanceof EsqlScalarFunction == false)) {
+                    if (supportsStringImplicitCasting(left.dataType())) {
+                        targetDataType = left.dataType();
+                        from = right;
+                    }
+                }
+                if (from != Literal.NULL) {
+                    Expression e = castStringLiteral(from, targetDataType);
```

Comment:
After removing the support to cast from string literals to datetime intervals, we don't need to check `e != from` in either `processBinaryOperator` or `processIn`, because when `castStringLiteral` is called, we know that a string literal will be cast to a target data type, and if the casting succeeded, we will get a `Literal` with target data type, otherwise we get an `UnresolvedAttribute`. 

Casting to datetime intervals is a little tricky with `Bucket`, as `buckets` accept both intervals and numeric values. I tried not changing `Bucket`'s second argument `buckets` to take a string type, it will require a `TransportVersion` to make it backward compatible. If users provide a valid interval string, we cast it to an interval successfully, if users provide an invalid interval string, we will return an `UnresolvedAttribute`. This means that `Bucket(float, "100")` will return a VerificationException, says cannot convert string 100 to either `DATE_PERIOD` or `TIME_DURATION`, we will not try to convert it to a numeric value.